### DISCUSSION TITLE
Feature "Шаблоны" - сохранения/загрузки записей в shared репозитории

### DISCUSSION THREAD

**Discussion by donvadicastro** [donvadicastro/case-management#21] [Ideas]
Мне кажется случайно затронув тему с репозиториями мы утянули тему в чуток неверное направление. Иметь что-то шаред подразумевает, что мы имеем точку пересечения доменов (где проэкт есть вершина графа этого домена). Я поэтому предлагаю не идти по этому пути, а сделать все сущности относительно проэкта только и не иметь ничего шаред.

Почему я думаю так правильно - в каком-то более глобальном абстрактном мире, где проэкты независимы- понятие shared отпадает само собой, потому как каждый проэкт менеджиться своими овнерами и они тоже независимы. Наличие shared подразумевает какое-то центральное управление, что в контексте 100+ проэктов уже не является возможным (ибо найти на 100 проэктов что-то общее для всех мне каежться невозможно).

Предлагаю пересмотреть этот критейрий.

**Comment by programp** [comment]
Shared между овнерами или, если конкретнее, между учетными записями - скорее всего лишнее, да. Но юзер этого приложения - это аналитик/продакт-менеджер, консолидирующий требования через, в частности, запись юзкейсов. Каждый такой аналитик имеет свой почерк и повторяющиеся от проекта приемы описания. Работая внесистемно, он каждый проект может начать описывать заново в новых терминах и категориях. Но системная работа дает возможность отточить подходы и шаблонизировать их. Рано или поздно аналитик видит каждый новый проект как повторение всех предыдущих с ответвлением от некой общей точки. 

Например, почти в каждом проекте будет набор actor/entity/action типа: 

**Actor:** 
- User Unregistered
- User Registered

**Entity:**
- User
- Account | Profile

**Action:**
- C.R.U.D 
- Get 
- Set 
- View
- Confirm

**...**

Большинство проектов действительно имеют общую внепредметную область - аккаунты, регистрации, базовые действия с сущностями и т.п. и это только маленький пример со списком того, что быстро пришло в голову. 

Иметь свой набор шаблонов, которые будут повторяться и кочевать из проекта в проект - это мастхэв для удобства и быстроты.
Напомню и повторюсь - основная задача, ради которой появилось это микро-приложение - это сократить путь от идеи до mvp. А он делится минимум на две части:
- **от модели к mvp**
   - здесь бустом может быть кодогенерация (полная или частичная);
- **от идеи к модели**
   - здесь бустят шаблонная система моделирования, которую мы уже сделали и можем совершенствовать, а также шаблоны проектных элементов внутри этой системы, о которых говорим сейчас.

Даже если посмотреть на эту проблему "от подхода", который, например, я культивирую, то это звучит так:
- чем шаблоннее будет каждый mvp ->
- тем быстрее они будут рождаться ->
- тем меньше времени будет потрачено на заблуждения в их пригодности (минимум) и единороговости (максимум) ->
- тем больше времени останется: 
   - либо на другие идеи, если эта будет непригодна; 
   - либо на доработку этой, если она окажется пригодной.

Поэтому даже принципиально - 
все, что есть "Шаблон" 
в руках человека, который всякие "Шаблоны" и "Паттерны" любит, 
а также умеет выделять любое общее из набора разного частного
и главное - делает из этих "Шаблонов" то, что должно делаться быстро
- это все можно и нужно.

Резюме. В такой постановке задачи шаблоны тут не то, что не лишние, а даже обязательные.

**Comment by donvadicastro** [comment]
OK, that makes sense. Тогда лишь стоит вынести это из списка требуемых на сейчас, поскольку в контексте первых 2 проэктов они не дадут особенного прироста производительности. По моему мнение они начинают играть роль, когда к-во проэктов растет и ты можешь быстро из каждого выбрать что-то, что перейдет в шаблон.

Посему я бы все равно немного переформатировал эту задачу со следующими изменениями:
* вынести в более высокий milestone (указать что она не блокирующая)
* иметь возможность существующие ресурсы (актеры, функции и т.д.) переводить в шаблоны, а не иметь несколько уровней управлениями ими (пер проэкт, глобал) - так мне кажеться будет удобнй (одним кликом существующий добавляешь в шаред, одним кликом из шаред его убираешь)

**Comment by programp** [comment]
Да, все, что не глобальное, то автоматом - проектное и никакого флага, кроме isShared у глобально-межпроектно-доступного, я не вижу и не видел. Может сформулировано было не точно. В конце, в последнем [сообщении](https://github.com/donvadicastro/case-management/issues/14#issuecomment-968274115) issue #14 я, кстати, показал еще один вариант реализации - без флага, а рефкой на глобальный репозиторий = если поле рефки у ресурса пустое, то значит ресурс проектный. 

Вот:
```
С точки зрения модели, это можно реализовать, добавив в сущность каждого из элементов модели свойство-рефку на один из соответствующих репозиториев. Например, какой-то конкретный Use Case теперь будет иметь рефку на репозиторий UC_SharedRepository_1, который может быть одним из нескольких (UC_SharedRepository_1, UC_SharedRepository_2, ...). Для каждого типа элемента модели (actor, action, use case, entity и тп.) будут свои репозитории.
```

**Comment by programp** [comment]
Объясню на пальцах.

**Есть таблица Repository:**

| Id  | type | name |
| - | - | - |
| 1  | UseCase  | UseCase_Rep_1 |
| 2 | Actor  | Actor_Rep_1 |
| 3 | Action  | Action_Rep_1 |
| 4 | Actor  | Actor_Rep_2 |

**Есть таблица Actor:**

| Id  | name | content | refRepository |
| - | - | - | - |
| 1 | Actor_1 | some content_1 | 2 |
| 2 | Actor_2 | some content_2 | 4 |
| 2 | Actor_3 | some content_3 |    |

Тут:
- Actor_1 - ссылается на репозиторий c Id равным 2, а значит входит репозиторий типа Actor с name : "Actor_Rep_1".
- Actor_2 - ссылается на репозиторий с id равным 4, а значит входит репозиторий типа Actor с name : "Actor_Rep_2".
- Actor_3 - не ссылается ни на какой репозиторий, а значит является исключительно проектным ресурсом.

**Reply by programp** [reply]
В такой модели можно легко получить: 
- все репозитории определенного типа (например, Actor) списком на странице создания/редактирования элемента модели Actor;
- при выборе конкретного репозитория из списка получить все ресурсы, с рефкой на него, и вывести их для возможности выбора (и загрузки в поля формы).

Я могу нарисовать, если нужно.
